### PR TITLE
ECS parent-child, isTrigger and input delaying

### PR DESCRIPTION
**Pull request by Peter-JanGootzen**
# Description
Creating a parent-child feature seemed easy, but it effected basically every single other system/feature. To prevent the massive scope-creep we were experiencing while building this PR we created [this document](https://docs.google.com/document/d/1nCuaUNFxBL2JVUqoIBVhkNmwr03OqiUPU62A5dUL8O0/edit?usp=sharing) to prevent ourselves from spending massive amounts of time on scenario's that would never come up in our game. Creating a full fledged parent-child system would have probably taken 3x as much time.

## Parent-Child
This PR centers around the parent-child feature in the EntityManager. An entity can now have a single parent and a parent can have multiple children. The transform of a child is always relative to that of it's parent, so when you need to know where an entity actually is you now need to use _getAbsoluteTransform_ in the EntityManager. When creating or setting the parent of an entity you can specify whether the current transform is already relative to its parent or not, this makes the system more usable and prevents bugs.

## CollisionDetectorInspector and isTrigger
To be able to detect that a collider is inside another collider, isTrigger was created by Jan. A collider now has the attribute isTrigger. The individual systems can then choose individually how to use the isTrigger. For instance the PhysicsSystem will not let an entity pass through an entity that has isTrigger and the PickupSystem will use isTrigger to detect whether the player is standing on a weapon.
Colliding or triggering with someone in your family hierarcy is impossible.

## Flipping
Before this PR we used to only have a flip in the Texture, this "flipping" has now been fully implemented into a directional feature. Every transform now has a xDirection and yDirection that can be positive or negative. The RenderingSystem will use the direction in Texture to flip the SDL_Texture accordingly.

The PhysicsComponent now has a flipX and flipY, these are used to determine what the PhysicsSystem should do when a flip should or could occur.

```
// If the entity is a parent, only the direction will get flipped
// If the entity is a child, both the direction and position get flipped
bool flipX;
bool flipY;
```

The position flipping relates to whether should have its position reversed when the parent gets flipped. A good example of this is the position of a weapon relative to its player that will get flipped when the player flips left or right.

# Kinematic
First we had `bool kinematic` now be have `Kinematic::SHOULD_NOT_BE_KINEMATIC`, beautiful is it not? From our features that needed to be supported in parent-child, we devised that a child should **always** be kinematic. This is something we try to enforce when a parent is set or when the child is thrown out the parents house. But to do this we had to remember what the original kinematic state was before the setting of a parent. So we now have:
```
// When using this enum in a system, use kinematic != IS_KINEMATIC or kinematic == IS_KINEMATIC
// to automaticly include or exclude the SHOULD_NOT_BE_KINEMATIC value
enum class Kinematic {
    IS_KINEMATIC,
    // This value should act the same as IS_KINEMATIC, however when a child is released
    // this value will get changed back to the original IS_NOT_KINAMTIC value
    SHOULD_NOT_BE_KINEMATIC,
    IS_NOT_KINEMATIC
};
```

# How can this be tested?

[Check the PR in BeastArena](https://github.com/Brick-Studios/BeastArena/pull/28).

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Included 1 or more screenshots
- [x] Code is const correct

### CODE REVIEW

File: include/brickengine/components/enums/kinematic.hpp [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#ifndef FILE_KINEMATIC_HPP
+#define FILE_KINEMATIC_HPP
+
+// When using this enum in a system, use kinematic != IS_KINEMATIC or kinematic == IS_KINEMATIC
+// to automaticly include or exclude the SHOULD_NOT_BE_KINEMATIC value
+enum class Kinematic {
+    IS_KINEMATIC,
+    // This value should act the same as IS_KINEMATIC, however when a child is released
+    // this value will get changed back to the original IS_NOT_KINAMTIC value
+    SHOULD_NOT_BE_KINEMATIC,
```
[the remark this answers is not in the document]
Comment (JanSchollaert) [reply include/brickengine/components/enums/kinematic.hpp]: You mean WAS_NOT_KINEMATIC?

File: include/brickengine/components/physics_component.hpp [hunk 2]
```diff
@@ -14,7 +16,11 @@ class PhysicsComponent : public ComponentImpl<PhysicsComponent> {
     double vx;
     double vy;
     bool gravity;
-    bool kinematic;
+    Kinematic kinematic;
+    // If the entity is a parent, only the direction will get flipped
+    // If the entity is a child, both the direction and position get flipped
+    bool flipX;
+    bool flipY;
```
Comment (Bram-Boris) [inline include/brickengine/components/physics_component.hpp]: The direction and position are both on the same axis. So shouldn't this be called. `flip_direction` and `flip_position`

File: include/brickengine/collision_detector.hpp [hunk 3]
```diff
@@ -40,12 +40,27 @@ Inspector Collision Detector
 #include "brickengine/enum/axis.hpp"
 #include "brickengine/enum/direction.hpp"
 
+struct CollisionReturnValues {
+    CollisionReturnValues(double spaceLeft, int objectId, bool isTrigger) : spaceLeft(spaceLeft), objectId(objectId), isTrigger(isTrigger) {}
```
Comment (Bram-Boris) [inline include/brickengine/collision_detector.hpp]: image
Any sneks in here?

File: include/brickengine/entities/entity_manager.hpp [hunk 4]
```diff
@@ -70,14 +81,97 @@ class EntityManager{
         components_by_class->at(componentType).insert_or_assign(entityId, std::move(component));
     }
 
-    void removeEntity(const int entityId){
+    void removeEntity(const int entityId) {
         for(auto& component : *components_by_class)
             component.second.erase(entityId);
     }
 
+    void setParent(int childId, int parentId, bool transformIsRelative) {
```
Comment (Bram-Boris) [inline include/brickengine/entities/entity_manager.hpp]: more snek 
image

File: include/brickengine/systems/physics_system.hpp [hunk 5]
```diff
@@ -10,6 +10,7 @@ class PhysicsSystem : public System {
 public:
     PhysicsSystem(std::shared_ptr<CollisionDetector> cd, std::shared_ptr<EntityManager> em);
     void update(double deltatime);
+    void updateChildren(int parentId);
```
Comment (Bram-Boris) [inline include/brickengine/systems/physics_system.hpp]: snek
image

File: lib/collision_detector.cpp [hunk 6]
```diff
@@ -8,117 +8,167 @@
...
-double CollisionDetector::spaceLeft(int entity, Axis axis, Direction direction) {
-    //We only support squares.
+CollisionReturnValues CollisionDetector::spaceLeft(int entity, Axis axis, Direction direction) {
+    // We only support rectangles
     auto entityRectCollider = entityManager->getComponent<RectangleColliderComponent>(entity);
-    auto entityTransform = entityManager->getComponent<TransformComponent>(entity);
+    auto [ entity_position, entity_scale ] = entityManager->getAbsoluteTransform(entity);
 
     auto collidableEntities = entityManager->getEntitiesByComponent<RectangleColliderComponent>();
+    auto children = entityManager->getChildren(entity);
```
Comment (Bram-Boris) [inline lib/collision_detector.cpp]: more snek
image

File: lib/collision_detector.cpp [hunk 7]
```diff
@@ -8,117 +8,167 @@
...
+                } 
+            }
+        }        
+    }    
+    return CollisionReturnValues(spaceLeft, objectId, isTrigger);
+}
+
+TriggerReturnValues CollisionDetector::isInTrigger(int entity){
+    // We only support rectangles
+    auto entityRectCollider = entityManager->getComponent<RectangleColliderComponent>(entity);
```
Comment (Bram-Boris) [inline lib/collision_detector.cpp]: your function, more snek
image

File: tests/entities/entity_manager_parent_child_test.cpp [hunk 8]
```diff
@@ -0,0 +1,55 @@
+#include <gtest/gtest.h>
+#include "brickengine/entities/entity_manager.hpp"
+
+#include "brickengine/components/component.hpp"
+#include "brickengine/components/transform_component.hpp"
+#include "brickengine/components/physics_component.hpp"
+#include "brickengine/components/player_component.hpp"
+
+TEST(EntityManager_Family, create_entity_with_parent_already_relative) {
+    EntityManager em;
```
Comment (Bram-Boris) [inline tests/entities/entity_manager_parent_child_test.cpp]: soo many snek :+1:

File: include/brickengine/components/physics_component.hpp (hunk 2 above)
Comment (Peter-JanGootzen) [reply include/brickengine/components/physics_component.hpp]: The _flipX_ and _flipY_ vars both dictate whether to flip direction and position on their corresponding axis. For instance when you would want to fly or swim up you could set _flipY_ to true and you weapon would flip upwards with you. This is something that we will most certainly never ever use, but it does make the purpose clear and easier to use.

Comment (Bram-Boris) [reply include/brickengine/components/physics_component.hpp]: Ah :+1:

**Pull request merged**